### PR TITLE
Fluent: migrate to CSS Modules

### PR DESCRIPTION
**Pull request by OEvgeny**
## Changelog Entry

(No CHANGELOG for tooling update)

## Description

This PR removes emotion from the `fluent-theme` package to:
- Reduce bundled size. The package went from ~70+ KB down to ~41 KB
- Reduce runtime complexity: emotion runs in runtime and while it's fast, not having it should be even faster
- Unlock native CSS authoring capabilities via CSS modules

## Design

For now we don't introduce any additional tools to the build pipeline. The `tsup` bundler uses `esbuild` which is able to bundle CSS modules when configured to use `local-css` loader. We now use this approach to bundle CSS into a single file.

After the file is produced, we're using `build.onEnd` hook to inject the styles as a string into the JavaScript bundle.

During script/module initialization, we inject the string into a style tag, and inject the tag into the `document.head` only if the document is available.

Styles are available internally, but not yet exposed in the package exports.

For types inference from the CSS modules, the `typescript-plugin-css-modules` was added.

## Specific Changes

- Move all styles to the corresponding `.module.css` files
- Add `injectCSSPlugin` esbuild plugin and make related changes
- Configure esbuild to bundle CSS modules
- Fix workspace typescript path
- Unprefix CSS rules, use power of CSS to namespace styles
  * Before `.webchat-fluent__sendbox__attachment-drop-zone`
  * After: `:global(.webchat-fluent) .sendbox__attachment-drop-zone`
- Replace previous `useStyles` hook with the new one handling CSS modules
- Add `*.module.css` type definitions
- Add `typescript-plugin-css-modules` TypeScript plugin

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] CSS styles reviewed (minimal rules, no `z-index`)
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -176,6 +176,7 @@
     "serve-handler": "^6.1.5",
     "strip-ansi": "^6.0.1",
     "typescript": "^5.3.2",
+    "typescript-plugin-css-modules": "^5.1.0",
```
Comment (OEvgeny) [inline package.json:179]: Had to add the package to the root dependencies as I discovered TypeScript tries to find it there

File: packages/fluent-theme/src/components/Theme.module.css [hunk 2]
```diff
@@ -0,0 +1,60 @@
+
+:global(.webchat-fluent).theme {
+  display: contents;
+
+  --webchat-colorNeutralForeground1: var(--colorNeutralForeground1, #242424);
```
Comment (compulim) [inline packages/fluent-theme/src/components/Theme.module.css:5]: Maybe this one later:

```suggestion
  --webchat--colorNeutralForeground1: var(--colorNeutralForeground1, #242424);
```

Or...

```suggestion
  --webchat-fluent--colorNeutralForeground1: var(--colorNeutralForeground1, #242424);
```

File: packages/fluent-theme/src/components/sendbox/ErrorMessage.module.css [hunk 3]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (compulim) [inline packages/fluent-theme/src/components/sendbox/ErrorMessage.module.css:1]: No empty new line here.

File: packages/fluent-theme/src/styles/useStyles.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { useMemo } from 'react';
+
+function useStyles<T extends CSSModuleClasses>(styles: T): T {
+  // @ts-expect-error: entries/fromEntries don't allow to specify keys type
+  return useMemo(
+    () =>
+      Object.freeze(
+        Object.fromEntries(Object.entries(styles).map(([baseCls, resultCls]) => [baseCls, `${baseCls} ${resultCls}`]))
```
Comment (compulim) [inline packages/fluent-theme/src/styles/useStyles.ts:8]:
```suggestion
        Object.fromEntries(Object.entries(styles).map(([baseClassName, resultClassName]) => [baseClassName, `${baseClassName} ${resultClassName}`]))
```

File: packages/fluent-theme/src/styles/injectStyle.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export const injectedStyles = '@--INJECTED-STYLES-CONTENT--@';
+
+export default function injectStyles() {
+  if (globalThis.document) {
+    const style = document.createElement('style');
```
Comment (compulim) [inline packages/fluent-theme/src/styles/injectStyle.ts:5]: We will need to add `nonce` attribute.

I expect a prop from the upstreamer.

```ts
root.render(
  <FluentThemeProvider nonce="a1b2c3d">
     ...
  </FluentThemeProvider>
);
```

And then we grab the nonce and inject it here.

You can do it in another PR.

File: packages/fluent-theme/tsup.config.ts [hunk 6]
```diff
@@ -21,10 +22,28 @@ const umdResolvePlugin = {
   }
 };
 
+const injectCSSPlugin = {
+  name: 'inject-css-plugin',
+  setup(build) {
+    build.onEnd(result => {
+      const js = result.outputFiles.find(f => f.path.match(/(\.js|\.mjs)$/u));
+      const css = result.outputFiles.find(f => f.path.match(/(\.css)$/u));
+      if (js && css && js.text.includes(injectedStylesPlaceholder)) {
```
Comment (compulim) [inline packages/fluent-theme/tsup.config.ts:31]:
```suggestion
      if (css && js?.text.includes(injectedStylesPlaceholder)) {
```

File: packages/fluent-theme/tsup.config.ts [hunk 7]
```diff
@@ -21,10 +22,28 @@ const umdResolvePlugin = {
...
+        const cssText = css.text;
+        js.contents = Buffer.from(js.text.replace(`"${injectedStylesPlaceholder}"`, JSON.stringify(cssText)));
```
Comment (compulim) [inline packages/fluent-theme/tsup.config.ts:33]:
```suggestion
        js.contents = Buffer.from(js.text.replace(`"${injectedStylesPlaceholder}"`, JSON.stringify(css.text)));
```

**Review by compulim — APPROVED** [review]
Approve with comments.

File: packages/fluent-theme/src/components/Theme.module.css (hunk 2 above)
Comment (OEvgeny) [reply packages/fluent-theme/src/components/Theme.module.css:5]: Let's address in the next PR

File: packages/fluent-theme/src/styles/injectStyle.ts (hunk 5 above)
Comment (OEvgeny) [reply packages/fluent-theme/src/styles/injectStyle.ts:5]: Sure, what should it do?

**Review by compulim — APPROVED** [review]
